server/rest/annotation.py

```diff
@@ -141,16 +144,62 @@ def getAnnotationSchema(self, params):
     )
     @access.cookie
     @access.public
-    @filtermodel(model='annotation', plugin='large_image')
     def getAnnotation(self, id, params):
         user = self.getCurrentUser()
-        annotation = Annotation().load(id, region=params, user=user, level=AccessType.READ)
+        annotation = Annotation().load(
+            id, region=params, user=user, level=AccessType.READ, getElements=False)
         if annotation is None:
             raise RestException('Annotation not found', 404)
         # Ensure that we have read access to the parent item.  We could fail
         # faster when there are permissions issues if we didn't load the
         # annotation elements before checking the item access permissions.
-        return annotation
+        #  This had been done via the filtermodel decorator, but that doesn't
+        # work with yielding the elements one at a time.
+        annotation = Annotation().filter(annotation, self.getCurrentUser())
+
+        annotation['annotation']['elements'] = []
+        breakStr = b'"elements": ['
+        base = json.dumps(annotation, sort_keys=True, allow_nan=False,
+                          cls=JsonEncoder).encode('utf8').split(breakStr)
+
+        def generateResult():
+            info = {}
+            idx = 0
+            yield base[0]
+            yield breakStr
+            for element in Annotationelement().yieldElements(annotation, params, info):
+                # The json conversion is fastest if we use defaults as much as
+                # possible.  The only value in an annotation element that needs
+                # special handling is the id, so cast that ourselves and then
+                # use a json encoder in the most compact form.
+                element['id'] = str(element['id'])
+                # Use ujson; it is much faster.  The standard json library
+                # could be used in its most default mode instead like so:
+                #   result = json.dumps(element, separators=(',', ':'))
+                result = ujson.dumps(element)
+                # By default, the json encoder outputs float values that happen
+                # to be integers as <value>.0.  This is a common occurrence.
+                # Javascript parses the json identically with and without the
+                # trailing '.0'.  If we can trivially do so, string these
+                # trailing values.  This splits the string on ", but only if it
+                # has no possibility of escaped quotes, then strips the
+                # trailing zeroes from the non-quoted sections, then puts the
+                # string back together.
+                if '.0' in result and '\\"' not in result:
+                    result = '"'.join([
+                        d if '.0' in d and di % 2 else d.replace('.0,', ',').replace('.0]', ']')
+                        for di, d in enumerate(result.split('"'))])
```

Comment:
This block makes me uncomfortable.  I would rather see the enhanced json serialization get moved to a function that gets unit tested independently.  To make it more readable, it would be better as a loop with better variable names.

I'm also curious about the performance implications of this.  Is the bandwidth saving worth the cost?  Would preprocessing the element dict to replace float's with int's be better than string manipulation?

Reply:
I'm not wholly convinced the bandwidth savings are worth the cost.  On a local machine, it is faster to not do this and accept that we push lots of useless trailing zeroes.  As soon as the machine isn't local (even just on the intranet), however, the bandwidth starts becoming the dominant factor.

Replacing floats with ints was substantially slower (I agree it was conceptually cleaner).  The json serialization code gets done in C if it uses defaults.  It is faster to use the defaults and the string replacement than either preprocessing floats to ints or overriding the serialization method.

I can just remove this, and if we ever decide the bandwidth is vital we can reintroduce it or another method.

Reply:
Sounds good to me.